scripts/opentitan/autoreg.py

```diff
@@ -0,0 +1,933 @@
+        }
+        reg_names -= noaccess_names
+        if not write and rc_names:
+            reg_names -= rc_names
+        rwidth = self._regwidth
+        dname = self._devname
+        hdname = dname.title().replace('_', '')
+        vnm = f'val{rwidth}'
+        if write:
+            if rwidth < 64:
+                vcast = f'uint{rwidth}_t val{rwidth} = (uint{rwidth}_t)val64;'
+            else:
+                vcast = ''
+            code = f'''
+            static void {dname}_regs_write(void *opaque, hwaddr addr,
+                                           uint64_t val64, unsigned size)
+            {{
+                {hdname}State *s = opaque;
+                (void)size;
+                {vcast}
+                hwaddr reg = R{rwidth}_OFF(addr);
+
+                uint32_t pc = ibex_get_current_pc();
+                trace_{dname}_io_write(s->ot_id, (uint32_t)addr, REG_NAME(reg),
+                                       val{rwidth}, pc);
+            '''
+        else:
+            if rwidth < 64:
+                vdef = f'uint{rwidth}_t val{rwidth};'
+            else:
+                vdef = ''
+            code = f'''
+            static uint64_t {dname}_regs_read(void *opaque, hwaddr addr,
+                                              unsigned size)
+            {{
+                {hdname}State *s = opaque;
+                (void)size;
+                {vdef}
+
+                hwaddr reg = R{rwidth}_OFF(addr);
+            '''
+        print(self.redent(code), file=tfp)
+        lines = []
+        lines.append('    switch (reg) {')
+        bitfield = (1 << rwidth) - 1
+        if not write:
+            for rname in sorted(reg_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+            lines.append(f'    {vnm} = s->regs[reg];')
+            lines.append('    break;')
+            for rname in sorted(rc_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+            lines.append(f'    {vnm} = s->regs[reg];')
+            lines.append('    s->regs[reg] = 0;')
+            lines.append('    break;')
+        else:
+            for rname in sorted(reg_names, key=lambda r: nregs[r].address):
+                lines.append(f'case R_{rname}:')
+                rbm, racc = reg_defs[rname]
+                if bitfield & rbm:
+                    lines.append(f'    {vnm} &= R_{rname}_WMASK;')
+                if racc in (QEMUAccess.RW1C, QEMUAccess.R0W1C):
+                    lines.append(f'    s->regs[reg] &= ~{vnm}; '
+                                 f'/* {racc.upper()} */')
+                elif racc == QEMUAccess.RW0C:
+                    lines.append(f'    s->regs[reg] &= {vnm}; /* RW0C */')
+                elif racc == QEMUAccess.RW1S:
+                    lines.append(f'    s->regs[reg] |= {vnm}; /* RW1S */')
+                else:
+                    lines.append(f'    s->regs[reg] = {vnm};')
```

Comment:
While I appreciate this tool doesn't support registers with multiple types of accesses on fields, maybe you could warn more explicitly in this case?

```suggestion
                elif racc == QEMUAccess.RW1S:
                    lines.append(f'    s->regs[reg] |= {vnm}; /* RW1S */')
                elif racc == QEMUAccess.UNDEF:
                    self._log.warning("Register '%s' has multiple access types which is "
                                      "not supported, defaulting to RW.", rname)
                    lines.append(f'    s->regs[reg] = {vnm};')
                else:
                    lines.append(f'    s->regs[reg] = {vnm};')
```

(I may have missed this being done elsewhere but couldn't see it).

Reply:
Sure, good idea

Reply:
Would you have an IP with such a combination of fields that I could use to check how the fix behaves? Thanks.

Reply:
Flash Controller, SPI Device & SPI Host all have a combination of `ro` and `rw1s` fields in INTR_STATE.

Reply:
Output with SPI_HOST:

```
 WARNING autoreg    Register 'INTR_STATE' has multiple access types which is not supported by this script, defaulting to RW.
    switch (reg) {
    case R_INTR_STATE:
        val32 &= R_INTR_STATE_WMASK;
        /* @todo handle multiple access type bifield */
        s->regs[reg] = val32;
        break;
    case R_INTR_TEST:
        s->regs[reg] = val32;
        break;
```